tests/baselines/reference/privateSymbolUnused.errors.txt

```diff
@@ -0,0 +1,39 @@
+tests/cases/conformance/classes/members/privateNames/privateSymbolUnused.ts(28,20): error TS6133: '[Symbol.asyncIterator]' is declared but its value is never read.
+tests/cases/conformance/classes/members/privateNames/privateSymbolUnused.ts(29,14): error TS6133: '[Symbol.iterator]' is declared but its value is never read.
+
+
+==== tests/cases/conformance/classes/members/privateNames/privateSymbolUnused.ts (2 errors) ====
+    // https://github.com/microsoft/TypeScript/issues/42051
+    export class Polling {
+        private async *[Symbol.asyncIterator]() {}
+        private *[Symbol.iterator]() {}
+    
+        async loop() {
+            for await (const updates of this) {
+            void updates
+            }
+        }
+        syncLoop() {
+            for (const updates of this) {
+            void updates
+            }
+        }
+    }
+    
+    
+    export class Spreadable {
+        private *[Symbol.iterator]() {}
+    
+        toArray() {
+            return [...this];
+        }
+    }
+    
+    export class Unused {
+        private async *[Symbol.asyncIterator]() {}
+                       ~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS6133: '[Symbol.asyncIterator]' is declared but its value is never read.
+        private *[Symbol.iterator]() {}
+                 ~~~~~~~~~~~~~~~~~
+!!! error TS6133: '[Symbol.iterator]' is declared but its value is never read.

```

Comment:
It’s a little confusing to consider what it means for these members to be marked as private, but let’s think about the consequences:

1. Within the program, consumers can spread or for/of an instance of `Unused` with no problems, but they can’t directly access this member via element access.
2. Outside the program (i.e., consumers of the declaration emit as a library) can spread or for/of an instance of `Unused`, but they don’t know what type it yields. (The member declaration is preserved, but the type annotation is erased.)

I think because these members are to some degree usable by external consumers, we can’t ever mark them unused (at least, not if the class is exported or globally accessible).

Reply:
> 2. Outside the program (i.e., consumers of the declaration emit as a library) can spread or for/of an instance of `Unused`, but they don’t know what type it yields. (The member declaration is preserved, but the type annotation is erased.)

🤔 Shouldn't symbol members behave like ordinary members? I don't know why not all following accesses report errors? [Playground](https://www.typescriptlang.org/play?noUnusedLocals=true&noUnusedParameters=true&target=99&ssl=1&ssc=1&pln=14&pc=1#code/MYGwhgzhAECqEFMAmB5ArgFwgSyQgwuFNAN4CwAUNNdAA4BO2AbmBgtAFQDaAygJ4BbAEYB7EADpsbeqxH0AugAoAlCWj0EGNPQB20AOT6A3NAC+lGnUYs20LvqFh6++dAC80Fe4B8pc1RoGZlZ2RwAvd09lHz9Kf0pgER0IDGgAMxERSJ0EAHc4RFRMHDxCSAgVI0oAemroADkshHp6OWhAZXJKDJFeQVEJKWbZBRUausboZtb6Dq7M+0dnJWUx6ABRFrbAUHI5kXFw0YogA)

```ts
class UsedOutsideClass {
    private *[Symbol.iterator](){ return ''; }
    private ['bar'] = () => {}
    private baz = () => {}
}

const foo = new UsedOutsideClass();
// No error ❓
foo[Symbol.iterator]()
// No error ❓
foo['bar']()
// Error ✅
foo.baz()
```

Reply:
Uhhh, that looks like a separate bug to me 😅

Reply:
The salient thing here isn’t the member declaration, it’s the access. `foo.bar` is an error but `foo['bar']` is not. Anyway, don’t worry about that for this PR.

Reply:
https://github.com/microsoft/TypeScript/issues/19335

Reply:
😕  So we should simply ignore private `Symbol.iterator/Symbol.asyncIterator` when checking unused members?

Reply:
Yes, at least when the class is global or exported.

Reply:
It occurs to me that maybe all well-known symbols should be ignored because they are designed to work implicitly and we have no practical way to track the usages.

```ts
class Foo {
  // Error: '[Symbol.toPrimitive]' is declared but its value is never read.
  private [Symbol.toPrimitive]() {
    return 1;
  }

  test() {
    // Logs 1
    console.log(String(this))
  }
}
```

Reply:
That seems plausible to me.